### PR TITLE
Inspq keycloak user module

### PR DESCRIPTION
**Pull request by elfelip**
##### SUMMARY
This PR add a new Keycloak module to manage users.

##### ISSUE TYPE
- New Module/Plugin Pull Request

##### COMPONENT NAME
keycloak_user

##### ADDITIONAL INFORMATION
This module include attributes and group membership management for users.

The alias username for the auth_username parameter have been removed because it conflicted with the username parameter of the Keycloak user representation

### CODE REVIEW

**Comment by elfelip** [comment]
Hello @eikef @mattock @ndclt
Someone can do something about this problem?
.github/BOTMETA.yml:0:0: Author elfelip not mentioned as active or inactive maintainer for plugins/modules/keycloak_user.py (mentioned are: eikef, ndclt, mattock)
Thankyou

**Comment by felixfontein** [comment]
> Someone can do something about this problem?
> .github/BOTMETA.yml:0:0: Author elfelip not mentioned as active or inactive maintainer for plugins/modules/keycloak_user.py (mentioned are: eikef, ndclt, mattock)
> Thankyou

Did you look at the part in CONTRIBUTING.md on adding new modules and plugins?

File: changelogs/fragments/6476-new-keycloak-user.module.yml
Comment (felixfontein) [inline changelogs/fragments/6476-new-keycloak-user.module.yml:2]: Please note that breaking changes are not accepted with a long deprecation period.

Instead of removing that alias for all modules, you should just remove that alias for the new module. In `DOCUMENTATION` you can specify
```
options:
    auth_username:
        aliases: []
```
to override it, and you will have to modify the return value of `keycloak_argument_spec()` in the module to remove that alias as well.

File: plugins/module_utils/identity/keycloak/keycloak.py
Comment (felixfontein) [inline plugins/module_utils/identity/keycloak/keycloak.py:766]: Is this a bugfix? Or does this change behavior of this existing function?

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:12]: Please remove `ANSIBLE_METADATA`. It has been deprecated for a long time and shouldn't be used in any new module or plugin.

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:20]:
```suggestion
short_description: Create and configure a user in Keycloak
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:22]:
```suggestion
    - This module creates, removes, or updates Keycloak users.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:32]: Paragraphs should always contain complete sentences.
```suggestion
            - Username for the user.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:37]:
```suggestion
            - ID of the user on the keycloak server if known.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:42]: Default values are better to be avoided if they refer to properties of the object that's modified. That way users can say "don't modify this value of an existing user when modifying it".

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:53]:
```suggestion
            - The user's first name.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:95]: Abbreviations should always be uppercase.
```suggestion
                    - Client ID of the client role. Not the technical ID of the client.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:220]:
```suggestion
    - The module does not modify the user ID of an existing user.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:218]: Always use FQCNs for stuff in community.general. Also add attributes
```suggestion
    - community.general.keycloak
    - community.general.attributes
attributes:
    check_mode:
        support: full
    diff_mode:
        support: full
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:227]: Always use FQCN.
```suggestion
      community.general.keycloak_user:
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:226]: Please dedent the examples so that they start in the first column.

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! I've added some first comments. Please note that you need to add tests as well (that's a requirement for every new module and plugin).

File: plugins/module_utils/identity/keycloak/keycloak.py
Comment (elfelip) [reply plugins/module_utils/identity/keycloak/keycloak.py:766]: In regards of the function documentation and it's usage in the keycloak_user_rolemapping module, this is a bug fix.
The original code returned an array of users instead of the user representation.
 def get_user_by_username(self, username, realm="master"):
        """ Fetch a keycloak user within a realm based on its username.

        If the user does not exist, None is returned.
        :param username: Username of the user to fetch.
        :param realm: Realm in which the user resides; default 'master'
        """

Comment (felixfontein) [reply plugins/module_utils/identity/keycloak/keycloak.py:766]: Could you please separate this bugfix into another PR so we can backport it? Bugfixes and features shouldn't be mixed in the same PR if possible.

File: changelogs/fragments/6476-new-keycloak-user.module.yml
Comment (felixfontein) [inline changelogs/fragments/6476-new-keycloak-user.module.yml:2]:
```suggestion
  - keycloak module utils - the function ``get_user_by_username`` now return the user representation or ``None`` as stated in the documentation (https://github.com/ansible-collections/community.general/pull/6476).
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:326]: Don't do this. Simply do `argument_spec['auth_username']['aliases'] = []`. This avoids potentially breaking other things the module_utils relies on.

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:23]:
```suggestion
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:291]:
```suggestion
    force: true
```
Always use `true`/`false` in YAML.

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:293]:
```suggestion
- name: Remove User
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:214]:
```suggestion
            - If C(true), allows to remove user and recreate it.
```

**Review by felixfontein — COMMENTED** [review]
Please note that in my earlier review you didn't look at all points. GitHub hides parts of the review if there are more then 10 comments, you need to explicitly un-hide them to see them.

File: plugins/module_utils/identity/keycloak/keycloak.py
Comment (felixfontein) [reply plugins/module_utils/identity/keycloak/keycloak.py:766]: Ping ^

Comment (elfelip) [reply plugins/module_utils/identity/keycloak/keycloak.py:766]: Here is the PR for this bug fix: https://github.com/ansible-collections/community.general/pull/6568

Comment (felixfontein) [reply plugins/module_utils/identity/keycloak/keycloak.py:766]: Can you rebase to the current `main` branch so this part disappears from this PR?

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:18]:
```suggestion
version_added: 7.1.0
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:98]:
```suggestion
                    - List of client roles to assign to the user.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:111]:
```suggestion
                    - Name of the group.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:115]:
```suggestion
                    - Control whether the user must be member of this group or not.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:128]:
```suggestion
                    - Credential type.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:133]:
```suggestion
                    - Value of the credential.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:138]:
```suggestion
                    - If C(true), the users are required to reset their credentials at next login.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:166]:
```suggestion
                    - Name of the attribute.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:170]:
```suggestion
                    - Values for the attribute as list.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:175]:
```suggestion
                    - Control whether the attribute must exists or not.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:204]:
```suggestion
            - Control whether the user should exists or not.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:305]:
```suggestion
  description: Message if it is the case.
```
If **what** is the case?

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:309]:
```suggestion
  description: Return C(true) if the operation changed the client on the keycloak server, C(false) otherwise.
```

File: changelogs/fragments/6476-new-keycloak-user.module.yml
Comment (felixfontein) [inline changelogs/fragments/6476-new-keycloak-user.module.yml:2]: Please remove this changelog fragment. New modules must not have changelog fragments.

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:34]:
```suggestion
            - ID of the user on the Keycloak server if known.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:42]:
```suggestion
            - Check the validity of user email.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:56]:
```suggestion
            - The user's last name.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:82]:
```suggestion
            - Client Authenticator Type.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:143]: I don't think this is possible to understand except if you already know what the option does.
```suggestion
            - RequiredActions user Auth.
```

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:151]:
```suggestion
            - List of IDPs of user.
```
I'm not sure this is correct. IdP = Identity Provider. But this seems to be more about identities, not about identity providers?

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:159]:
```suggestion
            - List of user attributes.
```
Please go through the remaining documentation and make sure that every description consists of complete sentences. Thanks.

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [inline plugins/modules/keycloak_user.py:335]: If *what* is the case?

File: .github/workflows/ansible-test.yml
Comment (felixfontein) [inline .github/workflows/ansible-test.yml:1]: Please restore this file.

File: plugins/modules/keycloak_user.py
Comment (felixfontein) [reply plugins/modules/keycloak_user.py:335]: I think this is the only comment left open. Can you please improve the description? Once that's fixed I'll merge.

**Review by felixfontein — APPROVED** [review]
I don't know Keycloak, but I think this looks good for a first release (at least the parts I can judge :) )! So if nobody objects, I'll merge this by the end of this week.

**Comment by felixfontein** [comment]
@elfelip thanks for your contribution!

**Pull request merged**
